src/__tests__/role.js

```diff
@@ -183,6 +184,138 @@ test('can include inaccessible roles', () => {
   expect(getByRole('list', {hidden: true})).not.toBeNull()
 })
 
+test('can be filtered by accessible name', () => {
+  const {getByRole} = renderIntoDocument(
+    `
+<div>
+  <h1>Order</h1>
+  <h2>Delivery Adress</h2>
+  <form aria-label="Delivery Adress">
+    <label>
+      <div>Street</div>
+      <input type="text" />
+    </label>
+    <input type="submit" />
+  </form>
+  <h2>Invoice Adress</h2>
+  <form aria-label="Invoice Adress">
+    <label>
+      <div>Street</div>
+      <input type="text" />
+    </label>
+    <input type="submit" />
+  </form>
+</div>`,
+  )
+
+  const deliveryForm = getByRole('form', {name: 'Delivery Adress'})
+  expect(deliveryForm).not.toBeNull()
+
+  expect(
+    // TODO: upstream bug in `aria-query`; should be `button` role
+    getQueriesForElement(deliveryForm).getByRole('textbox', {name: 'Submit'}),
+  ).not.toBeNull()
+
+  const invoiceForm = getByRole('form', {name: 'Delivery Adress'})
+  expect(invoiceForm).not.toBeNull()
+
+  expect(
+    getQueriesForElement(invoiceForm).getByRole('textbox', {name: 'Street'}),
+  ).not.toBeNull()
+})
+
+test('accessible name filter implements TextMatch', () => {
+  const {getByRole} = render(
+    `<h1>Sign <em>up</em></h1><h2>Details</h2><h2>Your Signature</h2>`,
+  )
+
+  // subset via regex
+  expect(getByRole('heading', {name: /gn u/})).not.toBeNull()
+  // regex
+  expect(getByRole('heading', {name: /^sign/i})).not.toBeNull()
+  // function
+  expect(
+    getByRole('heading', {
+      name: (name, element) => {
+        return element.nodeName === 'H2' && name === 'Your Signature'
+      },
+    }),
+  ).not.toBeNull()
+})
+
+test('includes accesible names in error message', () => {
+  const {getByRole} = render(`<h1>Sign <em>up</em></h1>`)
+
+  expect(() => getByRole('heading', {name: 'Sign Up'}))
+    .toThrowErrorMatchingInlineSnapshot(`
+"Unable to find an accessible element with the role "heading" and name "Sign Up"
+
+Here are the accessible roles:
+
+  heading:
+
+  Name "Sign up":
+  <h1 />
+
+  --------------------------------------------------
+
+<div>
+  <h1>
+    Sign 
+    <em>
+      up
+    </em>
+  </h1>
+</div>"
+`)
+
+  expect(() => getByRole('heading', {name: /Login/}))
```

Comment:
Not sure it's necessary to have all three of these because they're all testing the same codepath and their differences are covered in other tests.

Reply:
It's more about showcasing how the `name` argument is serialized. It may be too verbose in which case the codepath should change. And overall: Isn't that testing implementation details :wink:

Reply:
```suggestion
  expect(() => getByRole('heading', {name: /something that does not match/}))
```

I think this makes it more clear that it's not just a typo...